### PR TITLE
Factory/category{,Level,Item}

### PR DESCRIPTION
**Pull request by sorja**

### CODE REVIEW

File: src/category/category.test.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+  expect(categoryItem).toHaveProperty('uuid')
+  expect(categoryItem.uuid).toBeTruthy()
+
+  expect(categoryItem).toHaveProperty('props')
+  expect(categoryItem.props).toHaveProperty('code')
+  expect(categoryItem.props.code).toBe('001')
+  expect(categoryItem.props.labels).toBeTruthy()
+  expect(categoryItem.props.labels).toHaveProperty(LanguageCode.en)
+  expect(categoryItem.props.labels?.en).toBe('Label')
```
Comment (minotogna) [inline src/category/category.test.ts:63]: let's use mock object to verify values

File: src/category/factory.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
+import { Factory } from 'src/common'
+import { v4 as uuidv4 } from 'uuid'
+import { Category } from './category'
+import { CategoryItem, CategoryItemProps } from './item'
+import { CategoryLevel, CategoryLevelProps } from './level'
+
+export type CategoryFactoryParams = {
+  props?: {
```
Comment (minotogna) [inline src/category/factory.ts:8]: isn't this CategoryProps type?

File: src/category/factory.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  }
+  levels?: CategoryLevel[]
+  published?: boolean
+}
+
+export const CategoryFactory: Factory<Category> = {
+  createInstance: (params?: CategoryFactoryParams): Category => {
+    const defaultProps = {
+      props: {},
+      levels: [], // TODO: Use levels factory
```
Comment (minotogna) [inline src/category/factory.ts:19]: remove the TODO

File: src/category/factory.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    const { published, props, levels } = {
+      ...defaultProps,
+      ...params,
+    }
+
+    return {
+      levels,
+      props,
+      published,
+      uuid: uuidv4(),
```
Comment (minotogna) [inline src/category/factory.ts:31]: let's wrap uuid

File: src/category/factory.ts [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  index: number
+  props: CategoryLevelProps
+  categoryUuid: string
+}
+
+export const CategoryLevelFactory: Factory<CategoryLevel> = {
+  createInstance: (params?: CategoryLevelFactoryParams): CategoryLevel => {
+    const defaultProps = {
+      props: {},
+      index: undefined,
```
Comment (minotogna) [inline src/category/factory.ts:46]: default is 0. we shouldn't use undefined as default values

File: src/category/factory.ts [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    const { index, categoryUuid, props } = {
+      ...defaultProps,
+      ...params,
+    }
+
+    return {
+      categoryUuid,
+      index,
+      props,
+      uuid: uuidv4(),
```
Comment (minotogna) [inline src/category/factory.ts:58]: wrap uuid

File: src/category/factory.ts [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+  levelUuid: string
+  parentUuid?: string
+  props?: CategoryItemProps
+}
+
+export const CategoryItemFactory: Factory<CategoryItem> = {
+  createInstance: (params?: CategoryItemFactoryParams): CategoryItem => {
+    const defaultProps = {
+      props: {},
+      parentUuid: undefined,
```
Comment (minotogna) [inline src/category/factory.ts:73]: this is null. never use undefined

**Review by minotogna — CHANGES_REQUESTED** [review]

File: src/category/category.test.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    props: {
+      name: 'name',
+    },
+  }
+
+  const category = CategoryFactory.createInstance(categoryParams)
+
+  expect(category).toHaveProperty('props')
+  expect(category.props).toHaveProperty('name')
+  expect(category.props.name).toBe('name')
```
Comment (minotogna) [inline src/category/category.test.ts:15]: USE MOCK VALUES TO VERIFY VALUES!

File: src/category/category.test.ts [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+  expect(categoryLevel).toHaveProperty('categoryUuid')
+  expect(categoryLevel.categoryUuid).toBeTruthy()
+
+  expect(categoryLevel).toHaveProperty('index')
+  expect(categoryLevel.index).toBe(1)
+
+  expect(categoryLevel).toHaveProperty('props')
+  expect(categoryLevel.props).toHaveProperty('name')
+  expect(categoryLevel.props.name).toBeTruthy()
+})
```
Comment (minotogna) [inline src/category/category.test.ts:37]: USE MOCK VALUES TO VERIFY VALUES!

File: src/category/category.test.ts (hunk 1 above)
Comment (minotogna) [inline src/category/category.test.ts:63]: USE MOCK VALUES TO VERIFY VALUES!

File: src/category/factory.ts [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+  levels?: CategoryLevel[]
+  published?: boolean
+}
+
+export const CategoryFactory: Factory<Category> = {
+  createInstance: (params?: CategoryFactoryParams): Category => {
+    const defaultProps = {
+      props: {},
+      levels: [], // TODO: Use levels factory
+    }
```
Comment (minotogna) [inline src/category/factory.ts:20]: DO NOT ADD EMPTY DEFAULT VALUES

**Review by minotogna — CHANGES_REQUESTED** [review]

File: src/category/category.test.ts [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    props: {
+      name: 'name',
+    },
+  }
+
+  const category = CategoryFactory.createInstance(categoryParams)
+
+  expect(category).toHaveProperty('props')
+  expect(category.props).toHaveProperty('name')
+  expect(category.props.name).toBe(categoryParams.props?.name)
```
Comment (ramirobg94) [inline src/category/category.test.ts:22]:
```suggestion
expect(category).toHaveProperty('props')
  expect(category.props).toHaveProperty('name')
  expect(category.props.name).toBe(categoryParams.props?.name)

  expect(category.levels?.length).toBe(1)
  expect(category.levels?.[0].index).toBe(0)
  expect(category.levels?.[0].props.name).toBe('name')
```

File: src/category/factory.ts [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+
+    return {
+      levels,
+      props,
+      published,
+      uuid: UUIDs.v4(),
+    }
+  },
+}
```
Comment (ramirobg94) [inline src/category/factory.ts:32]:
```suggestion
export const CategoryFactory: Factory<Category> = {
  createInstance: (params?: CategoryFactoryParams): Category => {
    const defaultProps = {
      props: {},
    }

    const { published, props, levels } = {
      ...defaultProps,
      ...params,
    }

    const category = {
      levels,
      props,
      published,
      uuid: UUIDs.v4(),
    }
    return {
      ...category,
      levels: levels || [CategoryLevelFactory.createInstance({ ...category, categoryUuid: category.uuid })],
    }
  },
}
```

**Review by SteRiccio — DISMISSED** [review]

File: src/category/category.test.ts (hunk 8 above)
Comment (ramirobg94) [reply src/category/category.test.ts]: It was a bad merge, i think that now it is solved

**Comment by ramirobg94** [comment]
> see unresolved conversation

I think that it is solved now, it was a bad merge

File: src/category/category.ts [hunk 13]
```diff
@@ -15,13 +15,11 @@ export interface CategoryProps {
...
-  name: string
+  name?: string
 }
 
 export interface Category extends ArenaObject<CategoryProps> {
   id?: number
-  levels?: {
-    [key: number]: CategoryLevel
-  }
+  levels?: CategoryLevel[]
```
Comment (minotogna) [inline src/category/category.ts:23]: levels is an indexed object as before, not an array

**Review by minotogna — CHANGES_REQUESTED** [review]

File: src/category/category.ts (hunk 13 above)
Comment (ramirobg94) [reply src/category/category.ts:23]: // ====== CREATE
export const newCategory = (props = {}, levels = null) => {
  const category = {
    [keys.uuid]: uuidv4(),
    props,
  }
  return {
    ...category,
    [keys.levels]: levels || [newLevel(category)],
  }
}

Comment (ramirobg94) [reply src/category/category.ts:23]:
```
export const newLevel = (category, props = {}, index = R.pipe(getLevels, R.keys, R.length)(category)) => ({
  [CategoryLevel.keys.uuid]: uuidv4(),
  [CategoryLevel.keys.categoryUuid]: ObjectUtils.getUuid(category),
  [CategoryLevel.keys.index]: index,
  [CategoryLevel.keys.props]: {
    [CategoryLevel.keysProps.name]: `level_${index + 1}`,
    ...props,
  },
})```

**Review by minotogna — APPROVED** [review]

**Pull request merged**